### PR TITLE
[CPU] ScatterUpdate: dynamic shapes support

### PR DESCRIPTION
**Pull request by v-Golubev**
### Details:
 - *ScatterUpdate: dynamic shapes support*
 - *ScatterUpdate: added CPU tests with dynamic shapes*

### Tickets:
 - *70812*

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 1]
```diff
@@ -96,7 +92,7 @@ void MKLDNNScatterUpdateNode::initSupportedPrimitiveDescriptors() {
         IE_THROW() << errorPrefix << " should have same rank for input and output tensor";
     } else {
         for (size_t r = 0; r < srcRank; r++) {
-            if (srcDataDim[r] != dstDataDim[r]) {
+            if (srcDataDim[r] != Shape::UNDEFINED_DIM && dstDataDim[r] != Shape::UNDEFINED_DIM && srcDataDim[r] != dstDataDim[r]) {
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp:95]:
```suggestion
            if (!dimsEqualWeak(srcDataDim, dstDataDim[r])) {
```

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 2]
```diff
@@ -81,10 +77,10 @@ void MKLDNNScatterUpdateNode::initSupportedPrimitiveDescriptors() {
...
         return;
 
-    auto srcDataDim = getInputShapeAtPort(DATA_ID).getStaticDims();
-    auto indicesDim = getInputShapeAtPort(INDICES_ID).getStaticDims();
-    auto updateDim =  getInputShapeAtPort(UPDATE_ID).getStaticDims();
-    auto dstDataDim = getOutputShapeAtPort(0).getStaticDims();
+    auto srcDataDim = getInputShapeAtPort(DATA_ID).getDims();
+    auto indicesDim = getInputShapeAtPort(INDICES_ID).getDims();
+    auto updateDim =  getInputShapeAtPort(UPDATE_ID).getDims();
+    auto dstDataDim = getOutputShapeAtPort(0).getDims();
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp:83]:
```suggestion
    const auto& srcDataDim = getInputShapeAtPort(DATA_ID).getDims();
    const auto& indicesDim = getInputShapeAtPort(INDICES_ID).getDims();
    const auto& updateDim =  getInputShapeAtPort(UPDATE_ID).getDims();
    const auto& dstDataDim = getOutputShapeAtPort(0).getDims();
```

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 3]
```diff
@@ -111,26 +107,28 @@ void MKLDNNScatterUpdateNode::initSupportedPrimitiveDescriptors() {
...
+                    expectUpdateShape.push_back(indicesDim[ri]);
+                }
+                for (size_t rd = k; rd < srcRank; rd++) {
+                    expectUpdateShape.push_back(srcDataDim[rd]);
+                }
+                if (expectUpdateShape.size() != updateRank) {
+                    IE_THROW() << errorPrefix << " do not have matched tensor rank relationship for input, indices and update";
+                }
+                for (size_t ru = 0; ru < updateRank; ru++) {
+                    if (updateDim[ru] != Shape::UNDEFINED_DIM && expectUpdateShape[ru] != Shape::UNDEFINED_DIM && updateDim[ru] != expectUpdateShape[ru]) {
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp:129]: `dimsEqualWeak`
here and below

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 4]
```diff
@@ -182,7 +180,7 @@ void MKLDNNScatterUpdateNode::initSupportedPrimitiveDescriptors() {
             !getParentEdgeAt(DATA_ID)->getParent()->isConstant();
 
     NodeConfig config;
-    config.dynBatchSupport = false;
+    config.dynBatchSupport = true;
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp:183]: why changed?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 5]
```diff
@@ -182,7 +180,7 @@ void MKLDNNScatterUpdateNode::initSupportedPrimitiveDescriptors() {
             !getParentEdgeAt(DATA_ID)->getParent()->isConstant();
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp:180]: change previous line
add check on `!isDynamicNode()`

`bool canBeInplace = !isDynamicNode() && getParentEdgeAt(DATA_ID)->getParent()->getChildEdges().size() == 1 &&
            !getParentEdgeAt(DATA_ID)->getParent()->isConstant();`

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        ElementType idxPrecision;
+        std::tie(scatterParams, inputPrecision, idxPrecision) = this->GetParam();
+        const auto inputShapes = scatterParams.inputShapes;
+        const auto indicesValues = scatterParams.indicesValues;
+
+        init_input_shapes(inputShapes);
+
+        auto dataParams = ngraph::builder::makeDynamicParams(inputPrecision, { inputDynamicShapes[0], inputDynamicShapes[2] });
+        auto indicesParam = ngraph::builder::makeDynamicParams(idxPrecision, { inputDynamicShapes[1] });
+        auto scatter = std::make_shared<ngraph::opset4::ScatterNDUpdate>(dataParams[0], indicesParam[0], dataParams[1]);
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp:108]: please `set_friendly_name` for all params to be sure that they sorted in lexicographical order
same for other tests

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        ngraph::ParameterVector allParams{ dataParams[0], indicesParam[0], dataParams[1] };
+        function = std::make_shared<ngraph::Function>(ngraph::NodeVector{ scatter }, allParams, "ScatterNDUpdateLayerCPUTest");
+    }
+};
+
+TEST_P(ScatterNDUpdateLayerCPUTest, CompareWithRefs) {
+    SKIP_IF_CURRENT_TEST_IS_DISABLED()
+
+    run();
+    //CheckPluginRelatedResults(executableNetwork, "ScatterNDUpdate");
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp:119]: please don't forget to uncomment
same for other tests

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp (hunk 1 above)
Comment (v-Golubev) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: Fixed, thanks

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp (hunk 2 above)
Comment (v-Golubev) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: Corrected, thanks

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp (hunk 3 above)
Comment (v-Golubev) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: Fixed, thanks

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp (hunk 4 above)
Comment (v-Golubev) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: This change was reverted, thanks

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp (hunk 5 above)
Comment (v-Golubev) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: Corrected, thank you

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp (hunk 6 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp:113]: Done, thanks

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp (hunk 7 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp]: Done, thanks

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        result << inputPrecision << "_IS=";
+        for (const auto& shape : inputShapes) {
+            result << CommonTestUtils::partialShape2str({ shape.first }) << "_";
+        }
+        result << "TS=";
+        for (const auto& shape : inputShapes) {
+            result << "(";
+            if (!shape.second.empty()) {
+                auto itr = shape.second.begin();
+                do {
```
Comment (mandrono) [inline inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp:49]: minor
here and in other tests maybe better to use `for`?

**Review by mandrono — APPROVED** [review]

File: inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp (hunk 8 above)
Comment (v-Golubev) [reply inference-engine/tests/functional/plugin/cpu/single_layer_tests/scatter_ND_update.cpp]: Done, thanks!

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**